### PR TITLE
Fix unable to start web checkout in subsequent transactions

### PR DESCRIPTION
**Pull request by warmkesselj**
### Summary of changes

- This PR fixes an issue with allowing follow up transactions to occur. 
    - Provide a `removeObservers` method to clear out the observers on completion of a transaction.

 ### Checklist

 - [x] Added a changelog entry

### Authors
@warmkesselj 
@sshropshire 
@saperi22 

-

### CODE REVIEW

File: Demo/src/main/java/com/paypal/android/uishared/components/ActionButtonColumn.kt [hunk 1]
```diff
@@ -112,12 +113,15 @@ fun StatefulActionButtonPreview() {
...
-                    state = ActionState.Idle,
+                    state = ActionState.Success("Sample Value"),
                     onClick = {},
                     modifier = Modifier
                         .fillMaxWidth()
                         .padding(UIConstants.paddingMedium)
-                )
+                ) { state ->
+                    Text(text = "Sample Text", modifier = Modifier.padding(64.dp))
+
```
Comment (jaxdesmarais) [inline Demo/src/main/java/com/paypal/android/uishared/components/ActionButtonColumn.kt]:
```suggestion
```

File: Demo/src/main/java/com/paypal/android/uishared/components/ErrorView.kt [hunk 2]
```diff
@@ -22,3 +30,15 @@ fun ErrorView(error: Exception) {
         }
     }
 }
+
+
```
Comment (jaxdesmarais) [inline Demo/src/main/java/com/paypal/android/uishared/components/ErrorView.kt]:
```suggestion
```

File: Demo/src/main/java/com/paypal/android/uishared/components/ErrorView.kt [hunk 3]
```diff
@@ -22,3 +30,15 @@ fun ErrorView(error: Exception) {
...
+@Preview
+@Composable
+fun ErrorViewActionColumnPreview() {
+    MaterialTheme {
+        Surface(modifier = Modifier.fillMaxSize()) {
+                ErrorView(error = java.lang.Exception("Fake Exception"))
+
```
Comment (jaxdesmarais) [inline Demo/src/main/java/com/paypal/android/uishared/components/ErrorView.kt]:
```suggestion
```

File: CardPayments/src/main/java/com/paypal/android/cardpayments/CardClient.kt [hunk 4]
```diff
@@ -39,6 +41,8 @@ class CardClient internal constructor(
     private var approveOrderId: String? = null
     private val lifeCycleObserver = CardLifeCycleObserver(this)
 
+    private val activityRef = WeakReference(activity)
```
Comment (jaxdesmarais) [inline CardPayments/src/main/java/com/paypal/android/cardpayments/CardClient.kt]:
```suggestion
    private val activityReference = WeakReference(activity)
```

File: PayPalWebPayments/src/main/java/com/paypal/android/paypalwebpayments/PayPalWebCheckoutClient.kt [hunk 5]
```diff
@@ -41,13 +42,18 @@ class PayPalWebCheckoutClient internal constructor(
     var listener: PayPalWebCheckoutListener? = null
 
     /**
-     * Sets a listener to receive notificatioins when a Paypal Vault event occurs.
+     * Sets a listener to receive notifications when a Paypal Vault event occurs.
      */
     var vaultListener: PayPalWebVaultListener? = null
 
+    private val activityRef = WeakReference(activity)
```
Comment (jaxdesmarais) [inline PayPalWebPayments/src/main/java/com/paypal/android/paypalwebpayments/PayPalWebCheckoutClient.kt]:
```suggestion
    private val activityReference = WeakReference(activity)
```

**Review by jaxdesmarais — COMMENTED** [review]
It looks like there are some Detekt errors that we should resolve. Can we also add some tests around the `removeObservers` methods?

File: CardPayments/src/main/java/com/paypal/android/cardpayments/CardClient.kt [hunk 6]
```diff
@@ -197,4 +200,13 @@ class CardClient internal constructor(
         // TODO: consider either adding a listener method or next major version returning a result type
         cardVaultListener?.onVaultFailure(PayPalSDKError(1, "User Canceled"))
     }
+
+    /**
+     * Call this method at the end of the card flow
```
Comment (jaxdesmarais) [inline CardPayments/src/main/java/com/paypal/android/cardpayments/CardClient.kt]:
```suggestion
     * Call this method at the end of the card flow to clear out all observers and listeners
```

File: Demo/src/main/java/com/paypal/android/uishared/components/ActionButtonColumn.kt [hunk 7]
```diff
@@ -117,7 +118,9 @@ fun StatefulActionButtonPreview() {
                     modifier = Modifier
                         .fillMaxWidth()
                         .padding(UIConstants.paddingMedium)
-                )
+                ) { state ->
+                    Text(text = "Sample Text", modifier = Modifier.padding(64.dp))
```
Comment (jaxdesmarais) [inline Demo/src/main/java/com/paypal/android/uishared/components/ActionButtonColumn.kt:122]: Curious what this is needed for?

File: PayPalWebPayments/src/main/java/com/paypal/android/paypalwebpayments/PayPalWebCheckoutClient.kt [hunk 8]
```diff
@@ -119,4 +133,13 @@ class PayPalWebCheckoutClient internal constructor(
         analyticsService.sendAnalyticsEvent("paypal-web-payments:vault-wo-purchase:canceled")
         vaultListener?.onPayPalWebVaultCanceled()
     }
+
+    /**
+     * Call this method at the end of the web checkout flow
```
Comment (jaxdesmarais) [inline PayPalWebPayments/src/main/java/com/paypal/android/paypalwebpayments/PayPalWebCheckoutClient.kt]:
```suggestion
     * Call this method at the end of the web checkout flow to clear out all observers and listeners
```

File: CHANGELOG.md [hunk 9]
```diff
@@ -13,10 +13,13 @@
   * Add `returnUrl` property to `CardVaultRequest`
   * Add `authChallenge` property to `CardVaultResult`
   * Add `CardAuthChallenge` type
+  * Add `removeObservers` to `CardClient`
```
Comment (jaxdesmarais) [inline CHANGELOG.md]:
```suggestion
  * Add `CardClient.removeObservers()` method
```

File: CHANGELOG.md [hunk 10]
```diff
@@ -13,10 +13,13 @@
...
 * FraudDetection
   * Bump Magnes dependency to version 5.4.0
 * PaymentButtons
   * Update font typeface to "PayPalOpen" to meet brand guidelines
+* PayPalWebPayments
+  * Add `removeObservers` to `PayPalWebCheckoutClient`
```
Comment (jaxdesmarais) [inline CHANGELOG.md]:
```suggestion
  * Add `PayPalWebCheckoutClient.removeObservers` method
```

File: Demo/src/main/java/com/paypal/android/uishared/components/ActionButtonColumn.kt (hunk 7 above)
Comment (saperi22) [reply Demo/src/main/java/com/paypal/android/uishared/components/ActionButtonColumn.kt:122]: @jaxdesmarais looks like this is part of compose preview (annotated by @Preview) https://github.com/paypal/paypal-android/pull/251/files/cb5d5b7a7818424851e4a2f68406e5e9e233538b#diff-dbdf8e5b958c81bbcff72ecd0e61151a09aae144ce2f9726e2b1616511de536bR107 

It is a tool that allows users to view the UI in the IDE while developing.

Comment (warmkesselj) [reply Demo/src/main/java/com/paypal/android/uishared/components/ActionButtonColumn.kt:122]: This was just for previewing purposes to see the behavior when inflating the action column.

File: PayPalWebPayments/src/main/java/com/paypal/android/paypalwebpayments/PayPalWebCheckoutClient.kt [hunk 11]
```diff
@@ -13,7 +14,7 @@ import com.paypal.android.corepayments.analytics.AnalyticsService
  */
 class PayPalWebCheckoutClient internal constructor(
     // NEXT MAJOR VERSION: remove hardcoded activity reference
-    private val activity: FragmentActivity,
+    activity: FragmentActivity,
```
Comment (sarahkoop) [inline PayPalWebPayments/src/main/java/com/paypal/android/paypalwebpayments/PayPalWebCheckoutClient.kt]: Does this need to be public now?

Comment (saperi22) [reply PayPalWebPayments/src/main/java/com/paypal/android/paypalwebpayments/PayPalWebCheckoutClient.kt]: It is no longer a class variable - it's a linter optimization.
The scope of `activity` is now limited to the constructor and a class level variable, so the `activity` object is no longer needed after the constructor is done constructing. It is essentially a local object and needs no `private` modifier.

Comment (warmkesselj) [reply PayPalWebPayments/src/main/java/com/paypal/android/paypalwebpayments/PayPalWebCheckoutClient.kt]: Appears it can remain a private property. Pushing changes now.

File: PayPalWebPayments/src/main/java/com/paypal/android/paypalwebpayments/PayPalWebCheckoutClient.kt [hunk 12]
```diff
@@ -119,4 +133,13 @@ class PayPalWebCheckoutClient internal constructor(
         analyticsService.sendAnalyticsEvent("paypal-web-payments:vault-wo-purchase:canceled")
         vaultListener?.onPayPalWebVaultCanceled()
     }
+
+    /**
+     * Call this method at the end of the web checkout flow to clear out all observers and listeners
+     */
+    fun removeObservers() {
+        activityReference.get()?.let { it.lifecycle.removeObserver(observer) }
```
Comment (sarahkoop) [inline PayPalWebPayments/src/main/java/com/paypal/android/paypalwebpayments/PayPalWebCheckoutClient.kt:147]: is lifecycle always non null here?

**Comment by sarahkoop** [comment]
Looks like the latest google play compliance changes on main aren't updated here - can we update this branch to the latest main?

File: PayPalWebPayments/src/main/java/com/paypal/android/paypalwebpayments/PayPalWebCheckoutClient.kt (hunk 12 above)
Comment (saperi22) [reply PayPalWebPayments/src/main/java/com/paypal/android/paypalwebpayments/PayPalWebCheckoutClient.kt:147]: It is non-null all the time.

```
@NonNull
    @Override
    public Lifecycle getLifecycle() {
        return mLifecycleRegistry;
    }
```

Comment (warmkesselj) [reply PayPalWebPayments/src/main/java/com/paypal/android/paypalwebpayments/PayPalWebCheckoutClient.kt:147]: My guess would be that it's always not null, but I don't know that much about Android. Perhaps I can add some guards around it

File: PayPalWebPayments/src/main/java/com/paypal/android/paypalwebpayments/PayPalWebCheckoutClient.kt [hunk 13]
```diff
@@ -138,7 +138,10 @@ class PayPalWebCheckoutClient internal constructor(
      * Call this method at the end of the web checkout flow to clear out all observers and listeners
      */
     fun removeObservers() {
-        activityReference.get()?.let { it.lifecycle.removeObserver(observer) }
+        activity.lifecycle?.let { lifecycle ->
```
Comment (sarahkoop) [inline PayPalWebPayments/src/main/java/com/paypal/android/paypalwebpayments/PayPalWebCheckoutClient.kt]: Sorry I think we want to revert this change - it should be on the activityReference right? And I think Sai is correct it's always non-null, so disregard my comment

**Comment by sarahkoop** [comment]
+1 to Jax's comment - Is it possible to unit test removeObservers?

File: PayPalWebPayments/src/main/java/com/paypal/android/paypalwebpayments/PayPalWebCheckoutClient.kt (hunk 13 above)
Comment (saperi22) [reply PayPalWebPayments/src/main/java/com/paypal/android/paypalwebpayments/PayPalWebCheckoutClient.kt]: +1 need to revert the change.

File: Demo/src/main/java/com/paypal/android/uishared/components/ActionButtonColumn.kt (hunk 7 above)
Comment (jaxdesmarais) [reply Demo/src/main/java/com/paypal/android/uishared/components/ActionButtonColumn.kt:122]: Ah gotcha, didn't see it was in a preview and was confused by the "Sample Text". 😄

**Review by jaxdesmarais — APPROVED** [review]

File: PayPalWebPayments/src/main/java/com/paypal/android/paypalwebpayments/PayPalWebCheckoutClient.kt [hunk 14]
```diff
@@ -57,8 +62,11 @@ class PayPalWebCheckoutClient internal constructor(
      */
     fun start(request: PayPalWebCheckoutRequest) {
         analyticsService.sendAnalyticsEvent("paypal-web-payments:started", request.orderId)
-        payPalWebLauncher.launchPayPalWebCheckout(activity, request)?.let { launchError ->
-            notifyWebCheckoutFailure(launchError, request.orderId)
+
+        activityReference.get()?.let { activity ->
+            payPalWebLauncher.launchPayPalWebCheckout(activity, request)?.let { launchError ->
+                notifyWebCheckoutFailure(launchError, request.orderId)
+            }
```
Comment (sarahkoop) [inline PayPalWebPayments/src/main/java/com/paypal/android/paypalwebpayments/PayPalWebCheckoutClient.kt:69]: I think we should return an error if activity we can't get the activity reference - PayPal won't be launched but merchant wont get any notification (same for vault)

**Review by sarahkoop — APPROVED** [review]

File: PayPalWebPayments/src/main/java/com/paypal/android/paypalwebpayments/PayPalWebCheckoutClient.kt [hunk 15]
```diff
@@ -69,23 +81,32 @@ class PayPalWebCheckoutClient internal constructor(
...
-        payPalWebLauncher.deliverBrowserSwitchResult(activity)?.let { status ->
-            when (status) {
-                is PayPalWebStatus.CheckoutSuccess -> notifyWebCheckoutSuccess(status.result)
-                is PayPalWebStatus.CheckoutError -> status.run {
-                    notifyWebCheckoutFailure(error, orderId)
+        activityReference.get()?.let { activity ->
+            payPalWebLauncher.deliverBrowserSwitchResult(activity)?.let { status ->
+                when (status) {
+                    is PayPalWebStatus.CheckoutSuccess -> notifyWebCheckoutSuccess(status.result)
+                    is PayPalWebStatus.CheckoutError -> status.run {
```
Comment (tdchow) [inline PayPalWebPayments/src/main/java/com/paypal/android/paypalwebpayments/PayPalWebCheckoutClient.kt:100]: Is this `status.run {}` scope function needed?

File: detekt/detekt-config.yml [hunk 16]
```diff
@@ -145,7 +145,7 @@ complexity:
     thresholdInObjects: 11
     thresholdInEnums: 11
     ignoreDeprecated: false
-    ignorePrivate: false
+    ignorePrivate: true
```
Comment (tdchow) [inline detekt/detekt-config.yml:148]: Do we want to globally disable this detekt rule for private functions? I see some benefit in limiting the number of private functions as it can uncover a class that is handling too much.

Comment (sairaghukiran) [reply detekt/detekt-config.yml:148]: As long as we have the 600 or so line limit on each file, the number of functions shouldn't be a problem, IMO. I'm open to reverting the change and refactoring the class too. No strong opinion on the number of functions as long as the class itself is small.

File: PayPalWebPayments/src/main/java/com/paypal/android/paypalwebpayments/PayPalWebCheckoutClient.kt (hunk 15 above)
Comment (tdchow) [reply PayPalWebPayments/src/main/java/com/paypal/android/paypalwebpayments/PayPalWebCheckoutClient.kt:100]: Paired with Justin on this. This was existing code.

File: detekt/detekt-config.yml (hunk 16 above)
Comment (tdchow) [reply detekt/detekt-config.yml:148]: Paired with Justin on this - there wasn't a good way to refactor the private functions. Sai brings up good points. I'm good ignoring private functions as we have other checks in place.
